refactor(codegen): optimize binary receivers + fallback receiver special cases

anton-trunov opened the pull request:
## Issue

Closes #1946.
Closes #2357.

Supersedes #2358

Commit:
refactor(codegen): optimize binary receivers + fallback receiver special cases

Commit:
commit result json changes

Shvandre commented on src/generator/writers/writeRouter.ts:
```diff
@@ -288,6 +302,143 @@ function writeCommentReceivers(
+            1. x s′ −1
+            2. s 0
+
+            NEGATE DUP DUP
```
DUP DUP is equivalent to 2 0 BLKPUSH

Commit:
code review: DUP DUP -> 2 0 BLKPUSH

anton-trunov commented:
Throw `0xffff` for NotCoin

anton-trunov commented on src/generator/writers/writeRouter.ts:
```diff
@@ -453,8 +604,25 @@ function writeFallbackReceiver(
 ): void {
-    wCtx.append(`slice ${funcIdOf(fbRcv.selector.name)} = ${inMsg};`);
-    writeReceiverBody(fbRcv, contract, wCtx);
+    if (!isWildcard(fbRcv.selector.name) && fbRcv.ast.statements.length != 0) {
+        wCtx.append(`slice ${funcIdOf(fbRcv.selector.name)} = ${inMsg};`);
+    }
```
code review suggestion: use the kind of the fallback receiver

anton-trunov replied:
too many changes in unrelated places: there are more general cases where `writeFallbackReceiver` is used, let's just replicate a bit of this logic

anton-trunov commented on src/generator/writers/writeRouter.ts:
```diff
@@ -288,6 +302,143 @@ function writeCommentReceivers(
+    | { kind: "wildcard-parameter" }
+    | { kind: "statically-known-single-throw"; exitCode: number };
+
+function fallbackReceiverKind(
```
add comment: order is important!

anton-trunov commented on src/generator/writers/writeRouter.ts:
```diff
@@ -288,6 +302,143 @@ function writeCommentReceivers(
+                `;; Returns 32 bit message opcode, otherwise throws the "Invalid incoming message" exit code`,
+            );
+            wCtx.append(
+                ` ${loadOpcodeSignature} asm( -> 1 0) "32 LDUQ ${throwIfNot(exitCode)}";`,
```
```suggestion
                `${loadOpcodeSignature} asm( -> 1 0) "32 LDUQ ${throwIfNot(exitCode)}";`,
```

Commit:
PR number in results json

Kaladin13 committed:
add fallback throw tests

Commit:
notcoin throw + remove extra whitespace

Commit:
comment on precedence for fallbackReceiverKind

novusnota committed:
4Gb limits for tsc

it's not the first time tsc fails with OOM, especially when it
typechecks/transpiles tests
